### DISCUSSION TITLE
스코프가 뭔가요?

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#50] [스코프]
- 스코프가 뭔가요?
  - 스코프에는 어떤 종류가 있나요?
  - 스코프는 왜 필요한가요?
  - 🚧 (블록 스코프가 지역 스코프의 한 종류라고 한다면) 아래 실행 결과가 어떻게 될 지 설명해주세요.
    ```js
    var i = 10;
    
    for (var i = 0; i < 5; i++) {
      console.log(i);
    }
    
    console.log(i);
    ```
  - 블록 레벨 스코프와 함수 레벨 스코프의 차이점은 무엇인가요?
  - 스코프 측면에서 var와 let, const의 차이는?

**Comment by publdaze** [comment]
> - 스코프가 뭔가요?
- 식별자의 유효 범위
- 선언된 위치에 따라 유효 범위가 정해짐

>   - 스코프에는 어떤 종류가 있나요?
- 전역 스코프와 지역 스코프로 나눌 수 있음
- 전역 스코프에 선언된 전역 변수는 어디서든 참조 가능하고, 지역 스코프에 선언된 지역 변수는 자신의 지역 스코프와 하위 지역 스코프에서 유효

>  - 🚧 (블록 레벨 스코프가 지역 스코프의 한 종류라고 한다면) 아래 실행 결과가 어떻게 될 지 설명해주세요.
>     ```js
>     var i = 10;
>     
>     for (var i = 0; i < 5; i++) {
>       console.log(i);
>     }
>     
>     console.log(i);
>     ```
- 지역 스코프에 선언된 지역 변수가 자신의 스코프와 하위 스코프에서만 유효하기 때문에 for 문에 선언된 i는 for문 내부에서만 유효할 것 -> for문 외부인 스코프에서 console.log로 출력하고 있기 때문에 console.log에는 10이 찍힐 수 있음.
- 하지만 var의 경우는 함수 레벨 스코프를 따르기 때문에 for문의 코드 블록의 경우 지역 스코프로 잡힌다고 볼 수 없음 따라서 i의 최종 결과 값은 5가 됨

>  - 블록 레벨 스코프와 함수 레벨 스코프의 차이점은 무엇인가요?
- 블록 레벨 스코프는 모든 코드 블록이 지역 스코프로 인정되는 스코프를 의미
- 함수 레벨 스코프는 함수의 코드 블록만 지역 스코프로 인정되는 스코프를 의미
>  - 스코프 측면에서 var와 let, const의 차이는?
- var은 함수 레벨 스코프를 따르고, let과 const는 블록 레벨 스코프를 따름

**Comment by pipisebastian** [comment]
> 스코프가 뭔가요? 스코프는 왜 필요한가요? 스코프에는 어떤 종류가 있나요? 

- 식별자(변수, 함수명)의 유효범위입니다. 지역스코프, 전역스코프로 나뉩니다.
- 만약 스코프가 없다면, 같은 이름의 변수는 사용할 수 없게되며, 모든 변수의 유효범위를 체크할 수 없기에, 모두 전역변수가 되어버립니다.
- 전역/지역 스코프가 있습니다. 
  - 전역스코프는 코드의 가장 바깥 영역이며, 여기서 선언한 변수는 전역변수입니다. 
  - 지역스코프는 함수내부이며, 지역변수를 만듭니다.

> 아래 실행 결과가 어떻게 될 지 설명해주세요.

- `var`는 함수레벨 스코프를 따릅니다. 이에 for문에 선언되어있는 i의 경우, 블록레벨에 있기 때문에(함수레벨이 아님),  전역변수로 취급되어, i를 중복선언이 되기에, 자바스크립트 엔진은 이때, 2번째로 선언된 i의 키워드 var를 제외하고 동작하기 때문에, 전역변수 i를 덮어씌게 됩니다. 이에 출력은 0,1,2,3,4,5 입니다.

> 블록 레벨 스코프와 함수 레벨 스코프의 차이점은 무엇인가요?

- 함수레벨 스코프는, 함수에 의해서만 스코프가 생기는 반면, 블록 레벨 스코프는 블록 {} 단위로 스코프가 생깁니다. 

> 스코프 측면에서 var와 let, const의 차이는?

- 자바스크립트에서 키워드 `var` 의 경우, 함수레벨스코프를 따르나, `let`, `const` 는 블록레벨스코프를 따릅니다.
